### PR TITLE
[Feature] Adds instrumentation that remove function type annotations and renames function calls

### PR DESCRIPTION
**Pull request by Bobowski**
# Description

Adds new type of instrumentation: `REMOVE_ANNOTATIONS` and `FCALL_RENAME`
It's using php-parser to parse the code into AST and then remove all function type annotation (both for arguments and function return value) and then dumps the code to `*.php` again. For function call rename it works similarly but renames given function to the new name.

The main issue with this is altho it's far more robust than regex it's changing every file it goes through, meaning that executing it on any directory will recursively change every `.php` file inside (as backup the old file is kept with `*.php.preannotation` or `*.php.rerename` suffix)

Another issue is that it might impact applying patches / reverting patches as it naturally changes the structure of the file. Keep this in mind while preparing instrumentation!

`FCALL_RENAME` replaces `OVERRIDES` 

Example `instrumentation.ini`
```
PATCH_DIR="cakefuzzer/instrumentation/patches"
FCALL_RENAMES='
[
    {
        "path": "${APP_DIR}",
        "rename_from": "header",
        "rename_to": "__cakefuzzer_header"
    }
]
'
PATCHES='[]'
REMOVE_ANNOTATIONS='
[
    {
        "path": "${APP_DIR}"
    }
]
'
```

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] PR title contains the nature of changes (Docs, Feature, Fix, CI, etc.)
- [ ] I've set corresponding reviewers, set assignees, and labels.
- [ ] My code follows the style guidelines of this project (running `pre-commit`)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: cakefuzzer/phpfiles/instrumentation/install_php_parser.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+# Download the PHP-Parser library
+curl -L -o php-parser.zip https://github.com/nikic/PHP-Parser/archive/refs/heads/master.zip
```
Comment (dawid-czarnecki) [inline cakefuzzer/phpfiles/instrumentation/install_php_parser.sh:4]:
```suggestion
curl -Lso php-parser.zip https://github.com/nikic/PHP-Parser/archive/refs/heads/master.zip
```
To limit verbose mode.

**Review by dawid-czarnecki — CHANGES_REQUESTED** [review]
- Handle broken PHP files
```
Processing /var/www/cerebrate/src/../vendor/phpunit/php-code-coverage/tests/_files/Crash.php
PHP Fatal error:  Uncaught PhpParser\Error: Syntax error, unexpected '{', expecting T_STRING on line 2 in /cake_fuzzer/php-parser/lib/PhpParser/ParserAbstract.php:352
Stack trace:
#0 /cake_fuzzer/php-parser/lib/PhpParser/ParserAbstract.php(177): PhpParser\ParserAbstract->doParse()
#1 /cake_fuzzer/cakefuzzer/phpfiles/instrumentation/remove_annotations.php(58): PhpParser\ParserAbstract->parse()
#2 {main}
  thrown in /cake_fuzzer/php-parser/lib/PhpParser/ParserAbstract.php on line 352
```
- Add to `config/instrumentation_cake4.ini`:
```
REMOVE_ANNOTATIONS='
[
    {
        "path": "${APP_DIR}/../"
    }
]
```
- Modify/remove patches that remove annotations
- Modify/remove PATCHEs entries in `instrumentation.ini` files that remove annotations

**Review by dawid-czarnecki — CHANGES_REQUESTED** [review]
Apply patch that fixes PHP bug:
```
diff --git a/cakefuzzer/phpfiles/frameworks/cakephp/4/TargetAppHandler.php b/cakefuzzer/phpfiles/frameworks/cakephp/4/TargetAppHandler.php
index c385692..a68a10f 100644
--- a/cakefuzzer/phpfiles/frameworks/cakephp/4/TargetAppHandler.php
+++ b/cakefuzzer/phpfiles/frameworks/cakephp/4/TargetAppHandler.php
@@ -73,11 +73,9 @@ class TargetAppHandler extends CakePHPHandler {
 
         $dirs = Cake\Core\App::classPath('Controller');
         foreach($dirs as $dir) {
-            $paths = $this->_GetDirContents($dir);
+            $paths = $this->_searchFilesInDir($dir, $c_const.$ext, True);
             foreach($paths as $path) {
-                if(strpos($path, $c_const.$ext) !== false) {
-                    $controllers[] = substr($path, strlen($dir), -strlen($c_const)-strlen($ext));
-                }
+                $controllers[] = substr($path, strlen($dir), -strlen($c_const)-strlen($ext));
             }
         }
         $controllers = array_unique($controllers);
@@ -220,6 +218,27 @@ class TargetAppHandler extends CakePHPHandler {
         return $results;
     }
 
+    private function _searchFilesInDir($directory, $file_name, $ends_with = False) {
+        $dir = new RecursiveDirectoryIterator($directory);
+        $filter = new \RecursiveCallbackFilterIterator($dir, function ($current, $key, $iterator) use ($file_name, $ends_with) {
+            if (!$current->isDir()) {
+                if($ends_with) {
+                    if(strlen($current->getFilename()) <= strlen($file_name)) return False;
+                    return strpos(substr($current->getFilename(), -strlen($file_name)), $file_name) === 0;
+                }
+                return strpos($current->getFilename(), $file_name) === 0;
+            }
+            return True;
+        });
+        $iterator = new \RecursiveIteratorIterator($filter);
+        $files = array();
+        foreach ($iterator as $info) {
+            if(is_file($info->getPathname())) $files[] = $info->getPathname();
+        }
+        sort($files);
+        return $files;
+    }
+
     /**
      * Get arguments of a controller action
      * 
```

File: cakefuzzer/phpfiles/instrumentation/remove_annotations.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    echo "Error: Directory not found\n";
+    exit(1);
+}
+
+$directoryIterator = new RecursiveDirectoryIterator($directory);
+$iterator = new RecursiveIteratorIterator($directoryIterator);
+$phpFiles = new RegexIterator($iterator, '/\.php$/i', RecursiveRegexIterator::GET_MATCH);
+
+// Construct the iterator
+$it = new RecursiveDirectoryIterator($directory);
```
[the remark this answers is not in the document]
Comment (Bobowski) [reply cakefuzzer/phpfiles/instrumentation/remove_annotations.php:32]: yup. List might bloat the memory which is not the best here overall

**Comment by Bobowski** [comment]
> * Add annotation instruction to `instrumentation_cake4.ini`
> * Fix error on cerebrate instrumentation (remove in instrumentation.ini removed patches). Example:
> 
> ```
> cakefuzzer.instrumentation.InstrumentationError: Error: Error while instrumenting, got non-zero response from subprocess
> Try: patch /var/www/cerebrate/src/../vendor/laminas/laminas-diactoros/src/functions/normalize_server.php cakefuzzer/instrumentation/patches/CakePHP/4/vendor/laminas/laminas-diactoros/src/functions/normalize_server.php.patch
> ```
> 
> * After the instrumentation, I have to press enter to see the prompt. Last instrumentation command:
> 
> ```
> Type hints and annotations removed from '/var/www/cerebrate/src/../src/Command/ConsoleCommand.php', original file renamed to '/var/www/cerebrate/src/../src/Command/ConsoleCommand.php.preannotation'
> ```
> 
> * By running `check` after `apply` there sitll are some instrumentation to be done:
> 
> ```
> (venv) root@cerebrates:/cake_fuzzer# python3 cake_fuzzer.py instrument check
> Applied / Unapplied
> Overrides: 274/80
> Patches: 1/10
> Copies: 1/0
> Annotations: 0/1
> ```
> 
> * app_info `get_paths` is not working on this branch. Probably because changes from the main are not fully merged
> * Make sure that revert works properly. After I run revert I cannot access cerebrate from web. I get the error that suggest that patches were not reverted:
> 
> ```
> 2023-05-31 17:32:48 error: [RuntimeException] Identifier class `Fake` was not found. in /var/www/cerebrate/vendor/cakephp/authentication/src/Identifier/IdentifierCollection.php on line 106
> ```
> 
> * remove `.preannotation` files after revert

Are you sure you're using latest version ?
in `check` it should no longer show `Overrides` rather `FunctionCall Rename`

**Review by dawid-czarnecki — CHANGES_REQUESTED** [review]
After applying instrumentation and running check again it tells me that I'm still missing some instrumentation. I suspect it's because renames and annotations aren't known at this stage, and patches are incorrect because AST changed files but we need to find a way to present something meaningful.
```
...
Annotations Removed 1
(venv) root@cerebrates:/cake_fuzzer# python3 cake_fuzzer.py instrument check
Applied / Unapplied
FunctionCall Renames: 0/2
Patches: 2/9
Copies: 1/0
Annotations: 0/1
```

It doesn't support anonymous functions. Example: https://github.com/laminas/laminas-diactoros/blob/3.1.x/src/functions/normalize_uploaded_files.php#L33
It didn't remove annotations from anonymous functions here: ./vendor/laminas/laminas-diactoros/src/functions/marshal_uri_from_sapi.php

---

Replace: array_key_exists with __cakefuzzer_array_key_exists and remove patches
- ./vendor/laminas/laminas-diactoros/src/functions/marshal_uri_from_sapi.php
- ./vendor/cakephp/cakephp/src/Http/ServerRequest.php

Remove annotation changes in this patch: `./vendor/cakephp/cakephp/src/Http/ServerRequestFactory.php`

---

Make revert work. Right now it doesn't bring annotations back and doesn't revert patches.

---

Remove .preannotation files after revert

**Review by dawid-czarnecki — CHANGES_REQUESTED** [review]
.preannotation and .prerename files are still not removed
Reverts don't work:
```
(venv) root@cerebrates:/cake_fuzzer# python3 cake_fuzzer.py instrument revert
Annotations Reverted 0
FunctionCall Renames Reverted 0
Patches Reverted 2
Copies Reverted 1

(venv) root@cerebrates:/cake_fuzzer# python3 cake_fuzzer.py instrument check
Applied / Unapplied
FunctionCall Renames: x/3
Patches: 0/9
Copies: 0/1
Annotations: x/1

(venv) root@cerebrates:/cake_fuzzer# git log -n 1
commit 5372af24cc4fa985434619e08013a0a20aca453c (HEAD -> feature/instrumentation-remove-annotations, origin/feature/instrumentation-remove-annotations)
Merge: fc17b2b d79687a
Author: Adam Bobowski <bobowski.adam@gmail.com>
Date:   Fri Jun 9 20:14:15 2023 +0200

    Merge branch 'feature/instrumentation-remove-annotations' of github.com:Zigrin-Security/CakeFuzzer into feature/instrumentation-remove-annotations

root@cerebrates:/var/www/cerebrate# git diff vendor/wiremock-php/wiremock-php/test/WireMock/Serde/Type/SerdeTypePrimitiveTest.php | head -n 50 | tail
         $denormalized = $serdeType->denormalize($data, []);
         assertThat($denormalized, is($data));
     }
-
     /** @dataProvider providerMismatchedTypeAndData */
-    public function testDenormalizingMismatchedTypeDataThrows(string $type, $data)
+    public function testDenormalizingMismatchedTypeDataThrows($type, $data)
     {
         $this->expectExceptionMessage('Cannot deserialize data of type');
         $serdeType = new SerdeTypePrimitive($type);

```

**Review by dawid-czarnecki — APPROVED** [review]
Works!

**Pull request merged**